homeassistant/components/conversation/history.py

```diff
@@ -0,0 +1,333 @@
+    HomeAssistant,
+    callback,
+)
+from homeassistant.exceptions import HomeAssistantError, TemplateError
+from homeassistant.helpers import intent, llm, template
+from homeassistant.helpers.event import async_call_later
+from homeassistant.util import dt as dt_util, ulid
+from homeassistant.util.hass_dict import HassKey
+
+from .const import DOMAIN
+from .models import ConversationInput, ConversationResult
+
+DATA_CHAT_HISTORY: HassKey["dict[str, ChatHistory]"] = HassKey(
+    "conversation_chat_history"
+)
+DATA_CHAT_HISTORY_CLEANUP: HassKey["HistoryCleanup"] = HassKey(
+    "conversation_chat_history_cleanup"
+)
+
+LOGGER = logging.getLogger(__name__)
+CONVERSATION_TIMEOUT = timedelta(minutes=5)
+_NativeT = TypeVar("_NativeT")
+
+
+class HistoryCleanup:
+    """Helper to clean up the history."""
+
+    unsub: CALLBACK_TYPE | None = None
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the history cleanup."""
+        self.hass = hass
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self._on_hass_stop)
+        self.cleanup_job = HassJob(
+            self._cleanup, "conversation_history_cleanup", job_type=HassJobType.Callback
+        )
+
+    @callback
+    def schedule(self) -> None:
+        """Schedule the cleanup."""
+        if self.unsub:
+            return
+        self.unsub = async_call_later(
+            self.hass,
+            CONVERSATION_TIMEOUT.total_seconds() + 1,
+            self.cleanup_job,
+        )
+
+    @callback
+    def _on_hass_stop(self, event: Event) -> None:
+        """Cancel the cleanup on shutdown."""
+        if self.unsub:
+            self.unsub()
+        self.unsub = None
+
+    @callback
+    def _cleanup(self, now: datetime) -> None:
+        """Clean up the history and schedule follow-up if necessary."""
+        self.unsub = None
+        all_history = self.hass.data[DATA_CHAT_HISTORY]
+
+        # We mutate original object because current commands could be
+        # yielding history based on it.
+        for conversation_id, history in list(all_history.items()):
+            if history.last_updated + CONVERSATION_TIMEOUT < now:
+                del all_history[conversation_id]
+
+        # Still conversations left, check again in timeout time.
+        if all_history:
+            self.schedule()
+
+
+@asynccontextmanager
+async def async_get_chat_history(
+    hass: HomeAssistant,
+    user_input: ConversationInput,
+) -> AsyncGenerator["ChatHistory"]:
+    """Return chat history."""
+    all_history = hass.data.get(DATA_CHAT_HISTORY)
+    if all_history is None:
+        all_history = {}
+        hass.data[DATA_CHAT_HISTORY] = all_history
+        hass.data[DATA_CHAT_HISTORY_CLEANUP] = HistoryCleanup(hass)
+
+    history: ChatHistory | None = None
+
+    if user_input.conversation_id is None:
+        conversation_id = ulid.ulid_now()
+
+    elif history := all_history.get(user_input.conversation_id):
+        # Expire if it's an old conversation
```

Comment:
Could this be omitted if we always assume the the cleanup job has run and immediate update the last updated time? (or do it here and below). Seems nice if this didn't need to be considered in two places, assuming we can safely guard against a racing cleanup job. If that requires a mutex i can see why it may not be worth adding a mutex...

Reply:
oh good point. I forgot I had this code here, let me remove it.